OpenRA.Mods.Common/UtilityCommands/UpgradeRules.cs

```diff
@@ -1085,6 +1085,40 @@ internal static void UpgradeActorRules(int engineVersion, ref List<MiniYamlNode>
 					}
 				}
 
+				if (engineVersion < 20150607)
+				{
+					// Add WithRankDecoration to all actors using GainsExperience
+					var ge = node.Value.Nodes.FirstOrDefault(n => n.Key == "GainsExperience");
+					if (ge != null)
+					{
+						var nodeUpgrades = ge.Value.Nodes.FirstOrDefault(n => n.Key == "Upgrades");
+						var upgrades = nodeUpgrades != null ? nodeUpgrades.Value.Nodes.Count() : 4;
+
+						if (upgrades != 0)
+						{
+							foreach (var nodeUpgrade in nodeUpgrades.Value.Nodes)
+								nodeUpgrade.Value.Value = "rank" + (string.IsNullOrEmpty(nodeUpgrade.Value.Value) ? null : ", ") + nodeUpgrade.Value.Value;
+
+							var nodeChPal = ge.Value.Nodes.FirstOrDefault(n => n.Key == "ChevronPalette");
+							var chPal = nodeChPal != null && !string.IsNullOrEmpty(nodeChPal.Value.Value) ? nodeChPal.Value.Value : "effect";
+							ge.Value.Nodes.Remove(nodeChPal);
```

Comment:
Perhaps the `ChevronPalette` removal (these 3 lines) should not depend on `upgrades` being `!= 0` (which I find quite unlikely to happen, considering your fallback value of `4`), but instead be moved before the `if`.